Content/NPCs/Dungeon/CrescentCaster.cs

```diff
@@ -0,0 +1,505 @@
+        }
+
+        public List<Vector2> BezierPoints()
+        {
+            Vector2 startPoint = Owner.Center + new Vector2(Owner.spriteDirection * 3, -25);
+            BezierCurve curve = new BezierCurve(startPoint, MidPoint, TargetNPC.Center);
+            return curve.GetPoints(15);
+        }
+
+        public void Render(Effect effect)
+        {
+            trail?.Render(effect);
+            trail2?.Render(effect);
+        }
+    }
+
+    internal class CrescentCaster : ModNPC,IDrawPrimitive
+    {
+        public override string Texture => AssetDirectory.DungeonNPC + Name;
+
+        private const int XFRAMES = 3;
+        private readonly float ACCELERATION = 0.15f;
+        private readonly float MAXSPEED = 2;
+
+        private readonly int aiCounterReset = 800;
+
+        private int xFrame = 0;
+        private int yFrame = 0;
+
+        private int aiCounter = 0;
+
+        private List<NPC> supportTargets = new List<NPC>();
+
+        private List<CrescentCasterBolt> Bolts = new List<CrescentCasterBolt>();
+
+        private Player target => Main.player[NPC.target];
+
+        private bool Casting => aiCounter % aiCounterReset > 400;
+
+        private bool Supporting => xFrame == 2;
+
+        private float starOpacity = 0;
+
+        public override void Load()
+        {
+            for (int j = 1; j <= 5; j++)
+                GoreLoader.AddGoreFromTexture<SimpleModGore>(Mod, AssetDirectory.DungeonNPC + "CrescentCasterGore" + j);
+        }
+
+        public override void SetStaticDefaults()
+        {
+            DisplayName.SetDefault("Crescent Caster");
+            Main.npcFrameCount[NPC.type] = 10;
+
+            NPCID.Sets.NPCBestiaryDrawModifiers drawModifiers = new NPCID.Sets.NPCBestiaryDrawModifiers(0)
+            {
+
+            };
+            NPCID.Sets.NPCBestiaryDrawOffset.Add(Type, drawModifiers);
+        }
+
+        public override void SetDefaults()
+        {
+            NPC.width = 30;
+            NPC.height = 48;
+            NPC.damage = 0;
+            NPC.defense = 5;
+            NPC.lifeMax = 200;
+            NPC.value = 500f;
+            NPC.knockBackResist = 0.6f;
+            NPC.HitSound = SoundID.NPCHit2;
+            NPC.DeathSound = SoundID.NPCDeath2;
+        }
+
+        public override void SetBestiary(BestiaryDatabase database, BestiaryEntry bestiaryEntry)
+        {
+            bestiaryEntry.Info.AddRange(new IBestiaryInfoElement[]
+            {
+                BestiaryDatabaseNPCsPopulator.CommonTags.SpawnConditions.Biomes.TheDungeon,
+                new FlavorTextBestiaryInfoElement("[PH] He be castin")
+            });
+        }
+
+        public override void AI()
+        {
+            NPC.TargetClosest(true);
+            aiCounter++;
+
+            if (Casting)
+                CastBehavior();
+            else
+                WalkBehavior();
+        }
+
+        public override float SpawnChance(NPCSpawnInfo spawnInfo)
```

Comment:
Hm, I feel like the methods you call from AI should be defined after AI to keep behavior-related methods grouped together.

Reply:
fixed